src/live_restore/live_restore_server.c

```diff
@@ -0,0 +1,265 @@
+    __wt_spin_unlock(session, &server->queue_lock);
+}
+
+/*
+ * __live_restore_worker_run --
+ *     Entry function for a live restore thread. This is called repeatedly from the thread group
+ *     code so it does not need to loop itself.
+ */
+static int
+__live_restore_worker_run(WT_SESSION_IMPL *session, WT_THREAD *ctx)
+{
+    WT_DECL_RET;
+    WT_UNUSED(ctx);
+    WT_LIVE_RESTORE_SERVER *server = &S2C(session)->live_restore_server;
+
+    __wt_spin_lock(session, &server->queue_lock);
+    if (TAILQ_EMPTY(&server->work_queue)) {
+        /*
+         * Because we depend on stopping threads to avoid decrementing the counter indefinitely this
+         * section of code will both decrement the counter and terminate the thread. This is an
+         * unusual usage of thread groups but avoids refactoring a LOT of other code for the time
+         * being. The obvious fix would be to pass a WT_THREAD to the check function and then clear
+         * the run flag, however this then modifies the definition of __wt_cond_wait_signal which is
+         * used _everywhere_.
+         */
+        server->threads_working--;
+        if (server->threads_working == 0) {
+            /* All threads have finished their work, signal that source can be dropped. */
+            WT_STAT_CONN_SET(session, live_restore_state, WT_LIVE_RESTORE_COMPLETE);
+        }
+        /* Kill ourselves. */
+        F_CLR(ctx, WT_THREAD_RUN);
+        __wt_verbose_debug2(session, WT_VERB_FILEOPS, "Live restore worker terminating%s", "!");
+        __wt_spin_unlock(session, &server->queue_lock);
+        return (0);
+    }
+
+    WT_LIVE_RESTORE_WORK_ITEM *work_item = NULL;
+
+    work_item = TAILQ_FIRST(&server->work_queue);
+    WT_ASSERT(session, work_item != NULL);
+    TAILQ_REMOVE(&server->work_queue, work_item, q);
+    WT_STAT_CONN_SET(session, live_restore_queue_length, --server->queue_size);
+    __wt_verbose_debug2(
+      session, WT_VERB_FILEOPS, "Live restore worker taking queue item: %s", work_item->uri);
+    __wt_spin_unlock(session, &server->queue_lock);
+
+    WT_CURSOR *cursor;
+    WT_SESSION *wt_session = (WT_SESSION *)session;
+
+    /*
+     * TODO: Add logic to differentiate log files, we can't open cursors on them. Can they be
+     * deleted concurrently? It may be easier to have two separate queues.
+     */
+    /*
+     * Open a cursor so no one can get exclusive access on the object. This prevents concurrent
+     * schema operations like drop and rename.
+     */
+    WT_RET(wt_session->open_cursor(wt_session, work_item->uri, NULL, NULL, &cursor));
+
+    /*
+     * We need to get access to the WiredTiger file handle. Given we've opened the cursor we should
+     * be able to access the WT_FH by first getting to its block manager and then the WT_FH.
+     */
+    WT_ASSERT(session, (CUR2BT(cursor)->bm)->is_multi_handle == false);
+    /* This is a bit more than a little gross. */
+    WT_FILE_HANDLE *fh = CUR2BT(cursor)->bm->block->fh->handle;
```

Comment:
Little bit of layering knowledge going on here. Maybe have a local var for BT->bm. There is one use in block_mgr.c that does `bm->block->fh->handle` so block manager could have a "get_handle" method.

OR, there are several uses of `block->fh->handle` so maybe instead have a block->get_handle method.

Reply:
Created: https://jira.mongodb.org/browse/WT-13897